Tasks/Maven/mavenHelper.ps1

```diff
@@ -0,0 +1,152 @@
+		if (!$jdkPath) 
+		{
+			throw "Could not find JDK $jdkVersion $jdkArchitecture, please make sure the selected JDK is installed properly"
+		}
+
+		Write-Host "Setting JAVA_HOME to $jdkPath"
+		$env:JAVA_HOME = $jdkPath
+		Write-Verbose "JAVA_HOME set to $env:JAVA_HOME"
+	}
+}
+
+function PublishTestResults
+{
+	param([string]$publishJUnitResults,
+		  [string]$testResultsFiles)
+
+	$publishJUnitResultsFromAntBuild = Convert-String $publishJUnitResults Boolean
+
+	if($publishJUnitResultsFromAntBuild)
+	{
+	   # check for JUnit test result files
+		$matchingTestResultsFiles = Find-Files -SearchPattern $testResultsFiles
+		if (!$matchingTestResultsFiles)
+		{
+			Write-Host "No JUnit test results files were found matching pattern '$testResultsFiles', so publishing JUnit test results is being skipped."
+		}
+		else
+		{
+			Write-Verbose "Calling Publish-TestResults"
+			Publish-TestResults -TestRunner "JUnit" -TestResultsFiles $matchingTestResultsFiles -Context $distributedTaskContext
+		}    
+	}
+	else
+	{
+		Write-Verbose "Option to publish JUnit Test results produced by Maven build was not selected and is being skipped."
+	}
+}
+
+function RunSonarQubeAnalysis
+{
+	param([string]$sqAnalysisEnabled,
+		  [string]$sqConnectedServiceName,
+		  [string]$sqDbDetailsRequired,
+		  [string]$sqDbUrl,
+		  [string]$sqDbUsername,
+		  [string]$sqDbPassword, 
+		  [string]$userOptions,
+		  [string]$mavenPOMFile)
+
+	# SonarQube Analysis - there is a known issue with the SonarQube Maven plugin that the sonar:sonar goal should be run independently
+	$sqAnalysisEnabledBool = Convert-String $sqAnalysisEnabled Boolean
+
+	if ($sqAnalysisEnabledBool)
+	{
+		$sqServiceEndpoint = GetSonarQubeEndpointData $sqConnectedServiceName
+		$sqDbDetailsRequiredBool = Convert-String $sqDbDetailsRequired Boolean 
+
+		if ($sqDbDetailsRequiredBool)
+		{
+			$sqArguments = CreateSonarQubeArgs $sqServiceEndpoint.Url $sqServiceEndpoint.Authorization.Parameters.UserName $sqServiceEndpoint.Authorization.Parameters.Password $sqDbUrl $sqDbUsername $sqDbPassword
+		}
+		else
+		{
+			# The platform may cache the db details values so we force them to be empty
+			$sqArguments = CreateSonarQubeArgs $sqServiceEndpoint.Url $sqServiceEndpoint.Authorization.Parameters.UserName $sqServiceEndpoint.Authorization.Parameters.Password "" "" ""
+		}
+
+		$sqArguments = $userOptions + $sqArguments
+		Write-Verbose -Verbose "sonar:sonar options: $sqArguments"
+
+		Invoke-Maven -MavenPomFile $mavenPOMFile -Options $sqArguments -Goals "sonar:sonar"
+	 }
+}
+
+# Retrieves the url, username and password from the specified generic endpoint.
+# Only UserNamePassword authentication scheme is supported for SonarQube.
+function GetSonarQubeEndpointData
+{
+	param([string][ValidateNotNullOrEmpty()]$connectedServiceName)
+
+	$serviceEndpoint = Get-ServiceEndpoint -Context $distributedTaskContext -Name $connectedServiceName
```

Comment:
There is no equivalent for this in the nodejs API - Bryan is looking at it. Until then, we can simply add a warning that "SonarQube analysis is not yet enabled on the xplat agent"